pkg/operator/controller/route-metrics/controller.go

```diff
@@ -0,0 +1,180 @@
+// the resource.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log.Info("reconciling", "request", request)
+
+	// Fetch the Ingress Controller object.
+	ingressController := &operatorv1.IngressController{}
+	if err := r.client.Get(ctx, request.NamespacedName, ingressController); err != nil {
+		if kerrors.IsNotFound(err) {
+			// This means the Ingress Controller object was already deleted/finalized.
+			log.Info("Ingress Controller not found", "request", request)
+			// Delete the Shard label corresponding to the Ingress Controller from the RouteMetricsControllerRoutesPerShard metric.
+			DeleteRouteMetricsControllerRoutesPerShardMetric(request.Name)
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("failed to get Ingress Controller %q: %v", request, err)
+	}
+
+	// List all the Namespaces which matches the Namespace LabelSelector.
+	namespaceList := corev1.NamespaceList{}
+	if ingressController.Spec.NamespaceSelector != nil {
+		namespaceLabelSelector := ingressController.Spec.NamespaceSelector.MatchLabels
+		if err := r.client.List(ctx, &namespaceList, &client.ListOptions{
+			LabelSelector: labels.SelectorFromSet(namespaceLabelSelector),
+		}); err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to list Namespaces %q: %v", request, err)
+		}
+	} else {
+		// If Namespace LabelSelector is not provided then add Namespace with empty name to indicate all Namespaces.
+		namespaceList.Items = append(namespaceList.Items, corev1.Namespace{})
+	}
+
+	// Get the Route LabelSelector from the Ingress Controller.
+	var routeLabelSelector map[string]string
+	if ingressController.Spec.RouteSelector != nil {
+		routeLabelSelector = ingressController.Spec.RouteSelector.MatchLabels
+	}
+
+	// Store the previous value of the metric.
+	prevMetricValue := GetRouteMetricsControllerRoutesPerShardMetric(request.Name)
+
+	// Initialize the RouteMetricsControllerRoutesPerShard metric for Ingress Controller.
+	InitializeRouteMetricsControllerRoutesPerShardMetric(request.Name)
+
+	// Iterate through all the Namespaces.
+	for _, namespace := range namespaceList.Items {
+		// List all the Routes which matches the Route LabelSelector and the Namespace.
+		routeList := routev1.RouteList{}
+		if err := r.client.List(ctx, &routeList, &client.ListOptions{
+			LabelSelector: labels.SelectorFromSet(routeLabelSelector),
+			Namespace:     namespace.Name,
+		}); err != nil {
+			// Set the value of the metric to its previous value and re-queue the request.
+			SetRouteMetricsControllerRoutesPerShardMetric(request.Name, prevMetricValue)
+			return reconcile.Result{}, fmt.Errorf("failed to list Routes for the Shard %q: %v", request, err)
+		}
+
+		// Iterate through all the Routes.
+		for _, route := range routeList.Items {
+			// Check if the Route was admitted by the Ingress Controller.
+			if routeAdmitted(route, ingressController.Name) {
```

Comment:
Don't we still want to count it even it it's not admitted?  Then, you wouldn't have to iterate over the whole list, but could just count the number of routes and set the metric once.

Reply:
The routes which are not admitted by the Ingress Controller would not create any additional load on it as no traffic related to the routes would flow through it. Thus, counting only the routes which are admitted by the Ingress Controller will give the correct idea of the traffic load on it.

Reply:
Okay, but instead of calling `Inc` each time, just get the total and call `Set` one time instead.

Reply:
Okay, but instead of calling `Inc` each time, just get the total and call `Set` one time instead.  Something like:
```suggestion
			if routeAdmitted(route, ingressController.Name) {
			   counter++
			 }
		}
		SetRouteMetricsControllerRoutesPerShardMetric(ingressController.Name, counter)
```

Reply:
> The routes which are not admitted by the Ingress Controller would not create any additional load on it as no traffic related to the routes would flow through it. Thus, counting only the routes which are admitted by the Ingress Controller will give the correct idea of the traffic load on it.

Also, if the route isn't admitted, then OpenShift router will not write a backend for it to `haproxy.config`.  Thus the number of admitted routes impacts HAProxy's performance, and the number of non-admitted routes does not.  (Non-admitted routes will have an impact on OpenShift router as they'll be in its informer cache, but the impact should be negligible; the bigger concern is HAProxy.)